File: lab7/Dictionary.java
class Dictionary implements DictionaryInterface {
	private class Node{
		String key;
		String value;
		Node left;
		Node right;
		
		Node(String k, String v){
			key = k;
			value = v;
			left = null;
			right = null;
		}
	}
	
	private Node root;
	private int numItems;
	
	private Node findKey(Node R, String k){
		if(R == null || k.compareTo(R.key) == 0) return R;
		if(k.compareTo(R.key) < 0) return findKey(R.left, k);
		else return findKey(R.right, k);
	}
	
	private Node findParent(Node N, Node R){
			Node P = null;
			if(N != R){
			P = R;
			while(P.right != N && P.left != N){
				String nKey = N.key;
				if(nKey.compareTo(P.key) < 0) P = P.left;
				else P = P.right;
			}
			}
			return P;
			
	}
	
	private Node findLeftmost(Node R){
		Node L = R;
		if (L != null){
			for( ; L.left != null; L = L.left);
			}
		return L;
		
	}
	
	private void printInOrder(Node R){
		if(R != null){
			printInOrder(R.left);
			System.out.println(R.key + " " + R.value);
			printInOrder(R.right);
		}
	}
	
	private void deleteAll(Node N){
		if(N != null){
			deleteAll(N.left);
			deleteAll(N.right);
			N = null;
		}
	}
	
	public Dictionary(){
		root = null;
		numItems = 0;
	}
	
	public boolean isEmpty(){
		return(numItems == 0);
	}
	
	public int size(){
		return(numItems);
	}
	
	public String lookup(String k){
		Node N = findKey(root, k);
		if(N != null) return N.value;
		else return null;
	}
	
	public void insert(String k, String v) throws DuplicateKeyException{
		Node N = new Node(k, v);
		if(findKey(root, k) != null){
			throw new DuplicateKeyException("Dictionary error: insert() called with existing key");
		}
		Node P = null, F = root;
		while(F != null){
			P = F;
			if(F.key.compareToIgnoreCase(k) > 0) F = F.left;
			else F = F.right; 
		}
		
		if(P == null) root = N;
		else if(P.key.compareToIgnoreCase(k) > 0) P.left = N;
		else P.right = N;
		numItems++;
	}
	
	public void delete(String k) throws KeyNotFoundException{
		Node N = findKey(root, k);
		Node P, S;
		if(N == null){
			throw new KeyNotFoundException("Dictionary Error: cannot delete non-existing key");
		}
		if(N.left == null && N.right == null){
			if(N == root) root = null;
			else{
				P = findParent(N, root);
				if(P.right == N) P.right = null;
				else P.left = null;
			}
		}
		else if(N.right == null){
			if(N == root) root = N.left;
			else{
				P = findParent(N, root);
				if(P.right == N) P.right = N.left;
				else P.left = N.left;
			}
		}
		else if(N.left == null){
			if(N == root) root = N.right;
			else{
				P = findParent(N, root);
				if(P.right == N) P.right = N.right;
				else P.left = N.right;
			}
		}
		else{
			S = findLeftmost(N.right);
			N.key = S.key;
			N.value = S.value;
			P = findParent(S, N);
			if(P.right == S) P.right = S.right;
			else P.left = S.right;
		}
		numItems--;
	}
	
	public void makeEmpty(){
		deleteAll(root);
		root = null;
		numItems = 0;
	}
	
	public String toString(){
		printInOrder(root);
		return " ";
	}
	
}

File: pa5/Dictionary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

const int tableSize=101;

typedef struct NodeObj{
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

typedef NodeObj* Node;

Node newNode(char* k, char* v);

Node findKey(Dictionary D, char* key);

void freeNode(Node* pN);

Node newNode(char* k, char* v){
	Node N = malloc(sizeof(NodeObj));
	assert(N != NULL);
	N->key = k;
	N->value = v;
	N->next = NULL;
	return N;
}

void freeNode(Node* pN){
	if( pN != NULL && *pN != NULL){
		free(*pN);
		*pN = NULL;
	}
}



typedef struct DictionaryObj{
	Node* table;
	int numItems;
}DictionaryObj;


Dictionary newDictionary(){
	Dictionary D = malloc(sizeof(Dictionary));
	assert(D != NULL);
	D->table = calloc(tableSize, sizeof(Node));
	assert(D->table != NULL);
	D->numItems = 0;
	return D;
}

void freeDictionary(Dictionary* pD){
	if(pD != NULL && *pD != NULL){
		makeEmpty(*pD);
		free((*pD)->table);
		free(*pD);
		*pD = NULL;
	}
}



// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
 int sizeInBits = 8*sizeof(unsigned int);
 shift = shift & (sizeInBits - 1);
 if ( shift == 0 )
 return value;
 return (value << shift) | (value >> (sizeInBits - shift));
}
// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) {
 unsigned int result = 0xBAE86554;
 while (*input) {
 result ^= *input++;
 result = rotate_left(result, 5);
 }
 return result;
}
// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
 return pre_hash(key)%tableSize;
}


int isEmpty(Dictionary D){
	if(D == NULL){
		fprintf(stderr, "Dictionary Error: isEmpty() called on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	return(D->numItems == 0);
}

int size(Dictionary D){
	return(D->numItems);
}

Node findKey(Dictionary D, char* k){
	Node N = D->table[hash(k)];
	while(N != NULL){
		if(strcmp(N->key, k) == 0) return N;
		N = N->next;
	}
	return N;
}

char* lookup(Dictionary D, char* k){
	if(D == NULL){
		fprintf(stderr, "Dictionary Error: lookup() called on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	if(findKey(D, k) == NULL) return(NULL);
	Node N = findKey(D, k);
	return N->value;
}

void insert(Dictionary D, char* k, char* v){
	int index = hash(k);
	Node N = newNode(k, v);
	if(D == NULL){
		fprintf(stderr, "Dictionary Error: insert() called on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	if(findKey(D, k) != NULL){
		fprintf(stderr, "Dictionary Error: insert() called with existing key\n");
		exit(EXIT_FAILURE);
	}
	
	
	if(D->table[index] != NULL){
		N->next = D->table[index];
		D->table[index] = N;
	}
	else{
		D->table[index] = N;
	}
	D->numItems++;
}

void makeEmpty(Dictionary D){
	for(int i = 0; i < tableSize; i++){
		while(D->table[i] != NULL){
			Node N = D->table[i];
			D->table[i] = N->next;
			freeNode(&N);
			N = NULL;
		}
	}
	D->numItems = 0;
}

void delete(Dictionary D, char* k){
	int i = hash(k);
	Node P;
	Node N = D->table[i];
	
	if(findKey(D, k) == NULL){
		fprintf(stderr, "Dictionary Error: delete() called with non-existing key\n");
		exit(EXIT_FAILURE);
	}
	
	if(strcmp(N->key, k) == 0){
		D->table[i] = N->next;
		freeNode(&N);
		N = NULL;
	}
	else{
		P = D->table[i];
		N = findKey(D, k);
		while(P->next != N){
			P = P->next;
		}
		P->next = N->next;
		freeNode(&N);
		N = NULL;
	}
	D->numItems--;
}

void printDictionary(FILE* out, Dictionary D){
	Node N;
	for(int i = 0; i < tableSize; i++){
		N = D->table[i];
		while(N != NULL){
			fprintf(out, "%s %s\n", N->key, N->value);
			N = N->next;
		}
	}
}
